File: .agents/skills/tachyon-mcp/resources/java/ResourceHandlerExample.java
/*
 * Copyright (c) 2026 Konstantin Pavlov and contributors.
 */

import dev.tachyonmcp.server.domain.BlobResourceContents;
import dev.tachyonmcp.server.domain.TextResourceContents;
import dev.tachyonmcp.server.features.resources.AsyncResourceHandler;
import dev.tachyonmcp.server.features.resources.ResourceDescriptor;
import dev.tachyonmcp.server.features.resources.ResourceHandler;
import dev.tachyonmcp.server.features.resources.ResourceTemplateDescriptor;

import java.util.concurrent.CompletableFuture;

/**
 * Demonstrates static resources and URI-template resource registrations.
 */
final class ResourceHandlerExample {

    /**
     * Static resource — fixed URI.
     */
    static ResourceHandler configHandler() {
        return ResourceHandler.of((ctx, uri) ->
            TextResourceContents.of(uri, "{\"mode\":\"production\"}", "application/json"));
    }

    static ResourceDescriptor configDescriptor() {
        return ResourceDescriptor.of("server-config", "myapp://config", "Server configuration", "application/json");
    }

    /**
     * Static resource — binary (image, PDF, etc).
     */
    static ResourceHandler imageHandler() {
        return ResourceHandler.of((ctx, uri) ->
            BlobResourceContents.of(uri, "iVBORw0KGgoAAAANS...", "image/png"));
    }

    /**
     * URI template — {param} segments captured at runtime.
     */
    static ResourceTemplateDescriptor userProfileTemplateDescriptor() {
        return ResourceTemplateDescriptor.builder()
            .name("user-profile")
            .uriTemplate("myapp://users/{userId}/profile")
            .description("User profile data")
            .mimeType("application/json")
            .build();
    }

    static ResourceHandler userProfileTemplateHandler() {
        return (ctx, uri, params, uriTemplate) -> {
            var userId = params.get("userId").scalarValue();
            return TextResourceContents.of(uri, "{\"userId\":\"" + userId + "\"}", "application/json");
        };
    }

    /**
     * URI template — multi-segment with static prefix matching.
     */
    static ResourceTemplateDescriptor forecastTemplateDescriptor() {
        return ResourceTemplateDescriptor.of(
            "forecast", //name
            "weather://forecast/{city}" //uriTemplate
        );
    }

    static ResourceHandler forecastTemplateHandler() {
        return (ctx, uri, params, uriTemplate) -> TextResourceContents.of(
            uri, "{\"city\":\"" + params.get("city").scalarValue() + "\",\"temp\":22}", "application/json"
        );
    }

    /**
     * Async resource — returns a CompletionStage for non-blocking backends.
     * Blocking handlers run on virtual threads, so prefer plain ResourceHandler unless
     * integrating an already-async client.
     */
    static AsyncResourceHandler asyncConfigHandler() {
        return ResourceHandler.ofAsync((ctx, uri) -> CompletableFuture.supplyAsync(
            () -> TextResourceContents.of(uri, "{\"mode\":\"production\"}", "application/json")));
    }
}

File: .agents/skills/tachyon-mcp/resources/kotlin/ResourceHandlerExample.kt
// Copyright (c) 2026 Konstantin Pavlov and contributors.

package dev.tachyonmcp.skill

import dev.tachyonmcp.server.features.resources.ResourceDescriptor
import dev.tachyonmcp.server.kotlin.domain.Annotations
import dev.tachyonmcp.server.kotlin.domain.Icon
import dev.tachyonmcp.server.kotlin.domain.TextResourceContents
import dev.tachyonmcp.server.kotlin.features.resources.ResourceDescriptor
import dev.tachyonmcp.server.features.resources.ResourceHandler
import dev.tachyonmcp.server.features.resources.ResourceTemplateDescriptor
import dev.tachyonmcp.server.kotlin.features.resources.ResourceTemplateDescriptor
import dev.tachyonmcp.server.kotlin.features.resources.resourceDescriptor

/** Static resource — plain factory. */
fun configDescriptor(): ResourceDescriptor =
    ResourceDescriptor(
        name = "server-config",
        uri = "myapp://config",
        description = "Server configuration",
        mimeType = "application/json",
    )

/** DSL builder — all properties shown. */
fun richDescriptor(): ResourceDescriptor =
    resourceDescriptor(name = "user-profile", uri = "myapp://users/me") {
        description = "Current user profile"
        mimeType = "application/json"
        title = "My Profile"
        annotations = Annotations(priority = 0.5, lastModified = "2026-01-01T00:00:00Z")
        size = 1024
        icons =
            listOf(
                Icon(
                    src = "https://example.com/user.png",
                    mimeType = "image/png",
                    sizes = listOf("32x32"),
                    theme = "dark",
                ),
            )
    }

/** URI template — {param} segments captured at runtime. */
fun userProfileTemplateDescriptor(): ResourceTemplateDescriptor =
    ResourceTemplateDescriptor
        .builder()
        .name("user-profile")
        .uriTemplate("myapp://users/{userId}/profile")
        .description("User profile data")
        .mimeType("application/json")
        .build()

fun userProfileTemplateHandler(): ResourceHandler =
    ResourceHandler { _, request ->
        val userId = request.params()["userId"]?.scalarValue()
        TextResourceContents(
            uri = request.uri(),
            mimeType = "application/json",
            text = """{"userId":"$userId"}""",
        )
    }

/** URI template — multi-segment with static prefix matching. */
fun forecastTemplateDescriptor(): ResourceTemplateDescriptor =
    ResourceTemplateDescriptor
        .builder()
        .name("forecast")
        .uriTemplate("weather://forecast/{city}")
        .description("Weather forecast for a city")
        .mimeType("application/json")
        .build()

fun forecastTemplateHandler(): ResourceHandler =
    ResourceHandler { _, request ->
        val city = request.params()["city"]?.scalarValue()
        TextResourceContents(
            uri = request.uri(),
            mimeType = "application/json",
            text = """{"city":"$city","temp":22}""",
        )
    }
